lambda/src/simulated.rs

```diff
@@ -1,229 +1,94 @@
-    collections::VecDeque,
+    collections::HashMap,
     future::Future,
     io::Result as IoResult,
     pin::Pin,
     sync::{Arc, Mutex},
-    task::{Context, Poll, Waker},
+    task::{Context, Poll},
 };
-use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+use tokio::io::{AsyncRead, AsyncWrite, DuplexStream, ReadBuf};
 
-/// Creates a pair of `AsyncRead`/`AsyncWrite` data streams, where the write end of each member of the pair
-/// is the read end of the other member of the pair.  This allows us to emulate the behavior of a TcpStream
-/// but in-memory, deterministically, and with full control over failure injection.
-pub(crate) fn chan() -> (SimStream, SimStream) {
-    // Set up two reference-counted, lock-guarded byte VecDeques, one for each direction of the
-    // connection
-    let one = Arc::new(Mutex::new(BufferState::new()));
-    let two = Arc::new(Mutex::new(BufferState::new()));
-
-    // Use buf1 for the read-side of left, use buf2 for the write-side of left
-    let left = SimStream {
-        read: ReadHalf { buffer: one.clone() },
-        write: WriteHalf { buffer: two.clone() },
-    };
-
-    // Now swap the buffers for right
-    let right = SimStream {
-        read: ReadHalf { buffer: two },
-        write: WriteHalf { buffer: one },
-    };
-
-    (left, right)
-}
+use crate::Error;
 
 #[derive(Clone)]
 pub struct Connector {
-    pub inner: SimStream,
-}
-
-impl hyper::service::Service<Uri> for Connector {
-    type Response = SimStream;
-    type Error = std::io::Error;
-    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
-
-    fn poll_ready(&mut self, _: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        Poll::Ready(Ok(()))
-    }
-
-    fn call(&mut self, _: Uri) -> Self::Future {
-        let inner = self.inner.clone();
-        Box::pin(async move { Ok(inner) })
-    }
+    pub inner: Arc<Mutex<HashMap<Uri, DuplexStreamWrapper>>>,
 }
 
-impl Connection for SimStream {
-    fn connected(&self) -> hyper::client::connect::Connected {
-        hyper::client::connect::Connected::new()
-    }
-}
-
-/// A struct that implements AsyncRead + AsyncWrite (similarly to TcpStream) using in-memory
-/// bytes only.  Unfortunately tokio does not provide an operation that is the opposite of
-/// `tokio::io::split`, as that would negate the need for this struct.
-// TODO: Implement the ability to explicitly close a connection
-#[derive(Debug, Clone)]
-pub struct SimStream {
-    read: ReadHalf,
-    write: WriteHalf,
-}
-
-/// Delegates to the underlying `write` member's methods
-impl AsyncWrite for SimStream {
-    fn poll_write(mut self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &[u8]) -> Poll<IoResult<usize>> {
-        Pin::new(&mut self.write).poll_write(cx, buf)
-    }
-
-    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<IoResult<()>> {
-        Pin::new(&mut self.write).poll_flush(cx)
-    }
+pub struct DuplexStreamWrapper(pub DuplexStream);
```

Comment:
Let's not make this field public as public fields are potential forwards compatibility hazards.

```suggestion
pub struct DuplexStreamWrapper(DuplexStream);
```